Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs

```diff
@@ -0,0 +1,367 @@
+                    break;
+
+                case ReagentGrinderJuiceStartMessage msg:
+                    ClickSound();
+                    DoWork(user: message.Session.AttachedEntity!,isJuiceIntent:true);
+                    break;
+
+                case ReagentGrinderEjectChamberAllMessage msg:
+                    if(!ChamberEmpty)
+                    {
+                        ClickSound();
+                        for (var i = _chamber.ContainedEntities.Count - 1; i >= 0; i--)
+                        {
+                            EjectSolid(_chamber.ContainedEntities.ElementAt(i).Uid);
+                        }
+                    }
+                    break;
+
+                case ReagentGrinderEjectChamberContentMessage msg:
+                    if (!ChamberEmpty)
+                    {
+                        EjectSolid(msg.EntityID);
+                        ClickSound();
+                        _dirty = true;
+                    }
+                    break;
+
+                case ReagentGrinderEjectBeakerMessage msg:
+                    ClickSound();
+                    EjectBeaker(message.Session.AttachedEntity!);
+                    //EjectBeaker will dirty the UI for us, we don't have to do it explicitly here.
+                    break;
+            }
+        }
+        private void ClickSound()
+        {
+            _audioSystem.PlayFromEntity("/Audio/Machines/machine_switch.ogg", Owner, AudioParams.Default.WithVolume(-2f));
+        }
+        private void SetAppearance(ReagentGrinderVisualState state)
+        {
+            if (Owner.TryGetComponent(out AppearanceComponent? appearance))
+            {
+                appearance.SetData(PowerDeviceVisuals.VisualState, state);
+            }
+        }
+
+        public void OnUpdate()
+        {
+            if(_dirty)
+            {
+                UserInterface?.SetState(new ReagentGrinderInterfaceState
+                (
+                    _busy,
+                    HasBeaker,
+                    _chamber.ContainedEntities.Select(item => item.Uid).ToArray(),
+                    //Remember the beaker can be null!
+                    _heldBeaker?.Solution.Contents.ToArray()
+                ));
+                _dirty = false;
+            }
+        }
+
+        private void EjectSolid(EntityUid entityID)
+        {
+            if (_busy)
+                return;
+
+            if (_entityManager.EntityExists(entityID))
+            {
+                var entity = _entityManager.GetEntity(entityID);
+                _chamber.Remove(entity);
+
+                //Give the ejected entity a tiny bit of offset so each one is apparent in case of a big stack,
+                //but (hopefully) not enough to clip it through a solid (wall).
+                const float ejectOffset = 0.4f;
+                entity.Transform.LocalPosition += (_random.NextFloat() * ejectOffset, _random.NextFloat() * ejectOffset);
```

Comment:
Hmmm not sure if it should be LocalPosition or EntityCoordinates? @Zumorica probably knows more.